### PR TITLE
Adds the "Jestographic Sequencer", a clown-exclusive traitor item

### PR DESCRIPTION
**Pull request by FloFluoro**
## What Does This PR Do

jestographicsequencer

This PR adds the _**Jestographic Sequencer**_ (or cmag) to the clown's syndicate uplink. It's a 4TC card coated in a viscous slurry of electromagnetically charged bananium.

As you might have guessed from the name, its functionality is vaguely similar to its sister, the Cryptographic Sequencer; the clown can use it to gain access to restricted areas. Unlike the emag, however, the cmag doesn't force open airlocks or short out their circuitry. Instead, cmagging an airlock  **inverts its access**. For example, cmagging the bridge airlocks allows everyone _except_ members of command to enter, and cmagging a departmental door allows everyone _except_ members of that department through. This works the other way around; cmagging all-access doors converts them to _no_-access.

image
Inspecting a cmagged door will reveal that it's been coated in a gross, banana-scented sludge that impairs the airlock's ability to make rational decisions. If you attempt to pass through an airlock that you'd normally have access to, you will be denied, and the airlock will laugh at your misfortune. Trying to pass through a cmagged airlock by hiding your ID will result in the door mocking you for your transparent attempt at cheating the system.
image

To prevent those mean ol' heads and borgs from ruining the clown's clean, good-natured fun, cmagged doors don't respond to door remotes or silicon commands. They are otherwise fully functional... minus the inverted access.

"Oh god," I hear an Engineer main cry. "I hate emag man, and now I have to worry about cmag man?!" Fear not, because cmagging doesn't permanently damage the door! Once the access pad is cleaned with soap or space cleaner, the airlock will resume normal functionality.
image

To keep things manageable, the cmag currently only affects airlocks and windoors... but if this PR goes through, I plan on adding a _lot_ more functionality over time. Cmagging the cryotube to flood the mix with jestosterone? Cmagging the cloner so cloned crew emerge with the comic disability? Cmagging borgs to change their lawset to honksimov? The possibilities are as endless as they are horrible.

## Why It's Good For The Game
New traitor items are few and far between, and I'm of the opinion that SS13's lovable mascot deserves some TLC. The Jestographic Sequencer lets the clown commit the biggest prank of all, breaking and entering, in a unique way that's much funnier (and much less destructive) than emagging.

## Video of changes
https://user-images.githubusercontent.com/3611705/187347057-ae43c9c5-36d2-4160-a45b-2987d3a6b32a.mp4

## Testing
Compiled and tested on a local debug server. This is my first item and I poked my fingers in a lot of important stuff, so please excuse me if I did something horribly wrong.

## Changelog
:cl:
add: Adds the "Jestographic Sequencer", a new clown-exclusive traitor item
/:cl:

### CODE REVIEW

File: code/game/machinery/doors/door.dm [hunk 1]
```diff
@@ -227,6 +251,15 @@
 	return
 
 /obj/machinery/door/attackby(obj/item/I, mob/user, params)
+	if(cmagged) //Emags are Engineering's problem, cmags are the janitor's problem
+		if(istype(I, /obj/item/reagent_containers/spray/cleaner) || istype(I, /obj/item/soap))
```
Comment (lewcc) [inline code/game/machinery/doors/door.dm:255]: Maybe want to check if there are any reagents in the bottle?

File: code/game/machinery/doors/door.dm [hunk 2]
```diff
@@ -272,6 +305,36 @@
...
+	if(canopen)
+		if(ishuman(dooruser))
+			var/mob/living/carbon/human/H = dooruser
+			var/obj/item/card/id/idcard = H.get_idcard()
+			if(!idcard) //Humans can't game inverted access by taking their ID off.
+				if(density)
+					do_animate("deny")
+					if(hasPower())
+						to_chat(H, "<span class='warning'>The airlock speaker chuckles: 'What's wrong, pal? Lost your ID? Nyuk nyuk nyuk!'</span>")
+						playsound(src.loc, 'sound/machines/honkbot_evil_laugh.ogg', 25, TRUE, ignore_walls = FALSE)
```
Comment (lewcc) [inline code/game/machinery/doors/door.dm:326]: Implied src

```suggestion
						playsound(loc, 'sound/machines/honkbot_evil_laugh.ogg', 25, TRUE, ignore_walls = FALSE)
```

File: code/game/machinery/doors/door.dm [hunk 3]
```diff
@@ -272,6 +305,36 @@
...
+					return
+				return
+		if(density)
+			open()
+		else
+			close()
+	else
+		do_animate("deny")
+		if(hasPower())
+			playsound(src.loc, 'sound/machines/honkbot_evil_laugh.ogg', 25, TRUE, ignore_walls = FALSE)
```
Comment (lewcc) [inline code/game/machinery/doors/door.dm:336]:
```suggestion
			playsound(loc, 'sound/machines/honkbot_evil_laugh.ogg', 25, TRUE, ignore_walls = FALSE)
```

**Comment by datlo** [comment]
While this item is great, I feel like it should be 4 or 5 TC considering that it marks you, the clown, as almost definitely a traitor, bar the rarely used traitor crate. It's also a job specific job which imo should be slightly more cost efficient than regular ones.

**Comment by S34NW** [comment]
Should 100% be cheaper than 6TC

**Comment by datlo** [comment]
It'd be great if this item checks if the ID has at least a name and a job filled out, to prevent Captain and HOP bypassing the cmagged airlock with absolutely no effort using a spare blank ID. Not necessary, though.

**Comment by Komrad822** [comment]
This is really funny and all but what about clown Hijack. Yes the clown COULD just get 3 ID's with head status but they could just get an emag, or possibly forego both and get this item but with the negative of when they use it, it paints a massive target on their back. 

What im trying to say is, can it still be used to hijack the shuttle, whether it works like a REGULAR emag OR it removes the necessity of IDs needing to be head of staff IDs to force shuttle early leave (so like slip 2 assistants and steal their IDs).

File: code/game/machinery/doors/airlock.dm [hunk 4]
```diff
@@ -1378,6 +1380,18 @@ GLOBAL_LIST_EMPTY(airlock_emissive_underlays)
...
+		operating = TRUE
+		update_icon(AIRLOCK_EMAG, 1)
+		sleep(6)
+		if(QDELETED(src))
+			return
+		operating = FALSE
+		update_icon(AIRLOCK_CLOSED, 1)
+		cmagged = TRUE
+		return 1
+
```
Comment (Sirryan2002) [inline code/game/machinery/doors/airlock.dm:1394]:
```suggestion
/obj/machinery/door/airlock/cmag_act(mob/user)
	if(operating || cmagged || !density || !arePowerSystemsOn())
		return
	operating = TRUE
	update_icon(AIRLOCK_EMAG, 1)
	sleep(6)
	if(QDELETED(src))
		return
	operating = FALSE
	update_icon(AIRLOCK_CLOSED, 1)
	cmagged = TRUE
	return TRUE
```
early returns are preferred, could you double check the logic I recommended and make sure it's originally what you wanted?

File: code/game/machinery/doors/door.dm [hunk 5]
```diff
@@ -204,12 +216,24 @@
...
 		return TRUE
 	if(density)
+		if(cmagged)
+			cmag_switch(TRUE, user)
+			return
 		do_animate("deny")
+	else
+		if(cmagged)
+			cmag_switch(TRUE, user)
+			return
```
Comment (Sirryan2002) [inline code/game/machinery/doors/door.dm:236]:
```suggestion
	if(cmagged)
		cmag_switch(TRUE, user)
		return
	if(density)
		do_animate("deny")
```

File: code/game/machinery/doors/door.dm [hunk 6]
```diff
@@ -227,6 +251,25 @@
 	return
 
 /obj/machinery/door/attackby(obj/item/I, mob/user, params)
+	if(cmagged) //Emags are Engineering's problem, cmags are the janitor's problem
```
Comment (Sirryan2002) [inline code/game/machinery/doors/door.dm:254]: everything in this if statement needs to be extracted into its own proc

File: code/game/machinery/doors/door.dm [hunk 7]
```diff
@@ -227,6 +251,25 @@
...
+				cleaning = TRUE
+			else
+				return
+		if(istype(I, /obj/item/soap))
+			cleaning = TRUE
+
+		if(cleaning)
+			user.visible_message("<span class='notice'>[user] starts to clean the ooze off the access panel.</span>", "<span class='notice'>You start to clean the ooze off the access panel.</span>")
+			if(do_after(user, 50, target = src))
+				user.visible_message("<span class='notice'>[user] has cleaned the ooze off [src].</span>", "<span class='notice'>You've cleaned the ooze off [src].</span>")
```
Comment (Sirryan2002) [inline code/game/machinery/doors/door.dm:268]:
```suggestion
				user.visible_message("<span class='notice'>[user] cleans the ooze off [src].</span>", "<span class='notice'>You clean the ooze off [src].</span>")
```

File: code/game/machinery/doors/door.dm [hunk 8]
```diff
@@ -227,6 +251,25 @@
...
+				cmagged = FALSE
+				return
```
Comment (Sirryan2002) [inline code/game/machinery/doors/door.dm:270]:
```suggestion
```
redundant return

File: code/game/machinery/doors/door.dm [hunk 9]
```diff
@@ -272,6 +314,36 @@
 		emagged = TRUE
 		return TRUE
 
+/obj/machinery/door/cmag_act(mob/user)
+	if(density)
```
Comment (Sirryan2002) [inline code/game/machinery/doors/door.dm:318]: make this an early return please

File: code/game/machinery/doors/door.dm [hunk 10]
```diff
@@ -272,6 +314,36 @@
...
+		flick("door_spark", src)
+		sleep(6) //The cmag doesn't automatically open doors. It inverts access, not provides it!
+		cmagged = TRUE
+		return TRUE
+
+//Proc for inverting access on cmagged doors."canopen" should always return the OPPOSITE of the normal result.
+/obj/machinery/door/proc/cmag_switch(canopen, dooruser)
```
Comment (Sirryan2002) [inline code/game/machinery/doors/door.dm:325]: can you make `dooruser` just be `mob/living/user` for consistency sake

File: code/game/machinery/doors/door.dm [hunk 11]
```diff
@@ -272,6 +314,36 @@
...
+	if(canopen && !locked)
+		if(ishuman(dooruser))
+			var/mob/living/carbon/human/H = dooruser
+			var/obj/item/card/id/idcard = H.get_idcard()
+			if(!idcard) //Humans can't game inverted access by taking their ID off.
+				if(density)
+					do_animate("deny")
+					if(hasPower())
+						to_chat(H, "<span class='warning'>The airlock speaker chuckles: 'What's wrong, pal? Lost your ID? Nyuk nyuk nyuk!'</span>")
+						playsound(loc, 'sound/machines/honkbot_evil_laugh.ogg', 25, TRUE, ignore_walls = FALSE)
```
Comment (Sirryan2002) [inline code/game/machinery/doors/door.dm:335]: can you please test to see if this can be spammed?

File: code/game/machinery/doors/door.dm [hunk 12]
```diff
@@ -272,6 +314,36 @@
...
+//Proc for inverting access on cmagged doors."canopen" should always return the OPPOSITE of the normal result.
+/obj/machinery/door/proc/cmag_switch(canopen, dooruser)
+	if(canopen && !locked)
+		if(ishuman(dooruser))
+			var/mob/living/carbon/human/H = dooruser
+			var/obj/item/card/id/idcard = H.get_idcard()
+			if(!idcard) //Humans can't game inverted access by taking their ID off.
+				if(density)
+					do_animate("deny")
+					if(hasPower())
```
Comment (Sirryan2002) [inline code/game/machinery/doors/door.dm:333]: you should check for power before doing all this (Maybe L326?)

File: code/game/machinery/doors/windowdoor.dm [hunk 13]
```diff
@@ -247,6 +261,16 @@
 		open(2)
 		return 1
 
+/obj/machinery/door/window/cmag_act(mob/user, obj/weapon)
+	if(!operating && density && !cmagged)
```
Comment (Sirryan2002) [inline code/game/machinery/doors/windowdoor.dm:265]: early return this please

File: code/game/machinery/doors/windowdoor.dm [hunk 14]
```diff
@@ -247,6 +261,16 @@
...
+		cmagged = TRUE
+		operating = TRUE
+		flick("[base_state]spark", src)
+		playsound(src, "sparks", 75, TRUE, SHORT_RANGE_SOUND_EXTRARANGE)
+		sleep(6)
+		operating = FALSE
+		return 1
```
Comment (Sirryan2002) [inline code/game/machinery/doors/windowdoor.dm:272]:
```suggestion
		return TRUE
```

File: code/game/objects/items/weapons/cards_ids.dm [hunk 15]
```diff
@@ -80,6 +80,28 @@
...
+	AddComponent(/datum/component/slippery, src, 16 SECONDS, 100)
+
+/obj/item/card/cmag/attack()
+	return
+
+/obj/item/card/cmag/afterattack(atom/target, mob/user, proximity)
+	var/atom/A = target
+	if(!proximity)
+		return
+	A.cmag_act(user)
```
Comment (Sirryan2002) [inline code/game/objects/items/weapons/cards_ids.dm:103]:
```suggestion
	if(!proximity)
		return
	target.cmag_act(user)
```
no need to declare another var here, it's redundant, (unless this was done for a specific reason?)

**Review by Sirryan2002 — COMMENTED** [review]
first pass

File: code/game/machinery/doors/door.dm (hunk 11 above)
Comment (FloFluoro) [reply code/game/machinery/doors/door.dm]: I've added a one second cooldown that's handled in a new proc.

File: code/game/machinery/doors/door.dm [hunk 16]
```diff
@@ -226,7 +249,29 @@
...
+		return
+	user.visible_message("<span class='notice'>[user] starts to clean the ooze off the access panel.</span>", "<span class='notice'>You start to clean the ooze off the access panel.</span>")
+	if(do_after(user, 50, target = src))
+		user.visible_message("<span class='notice'>[user] cleans the ooze off [src].</span>", "<span class='notice'>You clean the ooze off [src].</span>")
+		cmagged = FALSE
+	return
+
 /obj/machinery/door/attackby(obj/item/I, mob/user, params)
+	if(cmagged)
+		clean_cmag_ooze(I, user)
```
Comment (Sirryan2002) [inline code/game/machinery/doors/door.dm:273]:
```suggestion
		clean_cmag_ooze(I, user)
		return
```
I believe you meant to do this? If not, just resolve and ignore this suggestion.

**Review by Sirryan2002 — COMMENTED** [review]
couple final things I noticed, after that I think this will be good code wise. Although, it would be nice to have another review take a second look at this.

File: code/game/machinery/doors/door.dm (hunk 16 above)
Comment (FloFluoro) [reply code/game/machinery/doors/door.dm:273]: Adding a return there would prevent this proc from checking for fireaxes on cmagged doors. I'll be adding a method to do `if(cmagged && I.canclean)` instead, which _would_ be able to return without issue, but that's out of scope of this PR.

**Comment by FloFluoro** [comment]
> It'd be great if this item checks if the ID has at least a name and a job filled out, to prevent Captain and HOP bypassing the cmagged airlock with absolutely no effort using a spare blank ID. Not necessary, though.

Done!

**Review by Sirryan2002 — APPROVED** [review]
code looks fine to me

File: code/game/objects/objs.dm [hunk 17]
```diff
@@ -34,6 +34,7 @@
 	var/suicidal_hands = FALSE // Does it requires you to hold it to commit suicide with it?
 	/// Is it emagged or not?
 	var/emagged = FALSE
+	var/cmagged = FALSE
```
Comment (hal9000PR) [inline code/game/objects/objs.dm:37]: make this a trait. I do not want a var on every single obj to track if its emagged by a clown.

**Review by hal9000PR — CHANGES_REQUESTED** [review]

File: code/game/machinery/doors/door.dm [hunk 18]
```diff
@@ -36,6 +36,9 @@
 	var/unres_sides = 0
 	//Multi-tile doors
 	var/width = 1
+	//Whether nonstandard door sounds (cmag laughter) are off cooldown.
+	var/soundready = TRUE
+	var/soundcooldown = 1 SECONDS
```
Comment (SteelSlayer) [inline code/game/machinery/doors/door.dm:41]: (Will have to be changed elsewhere too of course)
```suggestion
	var/sound_ready = TRUE
	var/sound_cooldown = 1 SECONDS
```

File: code/game/machinery/doors/door.dm [hunk 19]
```diff
@@ -272,6 +315,45 @@
...
+/obj/machinery/door/cmag_act(mob/user)
+	if(!density)
+		return
+	flick("door_spark", src)
+	sleep(6) //The cmag doesn't automatically open doors. It inverts access, not provides it!
+	cmagged = TRUE
+	return TRUE
+
+//Proc for inverting access on cmagged doors."canopen" should always return the OPPOSITE of the normal result.
+/obj/machinery/door/proc/cmag_switch(canopen, mob/living/user)
```
Comment (SteelSlayer) [inline code/game/machinery/doors/door.dm:327]: Can you make this proc use early returns?

File: code/game/machinery/doors/door.dm [hunk 20]
```diff
@@ -226,7 +249,28 @@
 /obj/machinery/door/proc/try_to_crowbar(mob/user, obj/item/I)
 	return
 
+/obj/machinery/door/proc/clean_cmag_ooze(obj/item/I, mob/user) //Emags are Engineering's problem, cmags are the janitor's problem
```
Comment (SteelSlayer) [inline code/game/machinery/doors/door.dm:252]: So this is.. alright for the moment, but I don't want this turning into attackby 2 with a big amount of ifs for cleaning items. In the future it should be handled on the attacking item (soap, etc.), and use `clean_act` or something like we were all talking about on discord the other day.

File: code/game/objects/objs.dm (hunk 17 above)
Comment (FloFluoro) [reply code/game/objects/objs.dm]: Done.

File: code/game/machinery/doors/door.dm (hunk 18 above)
Comment (FloFluoro) [reply code/game/machinery/doors/door.dm]: Done.

File: code/game/machinery/doors/door.dm (hunk 19 above)
Comment (FloFluoro) [reply code/game/machinery/doors/door.dm:327]: Done.

File: code/game/machinery/doors/door.dm (hunk 20 above)
Comment (FloFluoro) [reply code/game/machinery/doors/door.dm:252]: Will be handling this in a future PR that standardizes cleaning stuff, thanks!

**Review by hal9000PR — APPROVED** [review]

**Review by SteelSlayer — APPROVED** [review]

**Pull request merged**